[jit] fix @property

suo opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#28395 [jit] fix @property**

Currently `@property` methods are broken in TorchScript because we
basically treat it as an attribute in the existing path: we'll evaluate
the method once and store that as the value forever.

Since lack of `@property` support is easily worked around (just make it
a method), I've opted to just explicitly error to avoid confusion. If
people want it, they can file an issue and we can look at their use
case.

This also helps us nicely clean up some parts of the ScriptModule conversion
path.

Pull Request resolved: https://github.com/pytorch/pytorch/pull/28395

Differential Revision: [D18054946](https://our.internmc.facebook.com/intern/diff/D18054946)

suo committed:
[jit] fix @property

Currently `@property` methods are broken in TorchScript because we
basically treat it as an attribute in the existing path: we'll evaluate
the method once and store that as the value forever.

This PR fixes it so that `@property` actually behaves correctly:
1. Added `properties` as part of `ConcreteModuleType`.
2. Added a `PropertyValue` `SugaredValue` that behaves exactly like a
method but it can be used as a value.
3. `ModuleValue::attr` now checks wheter a method is a property, and
returns a `PropertyValue` if so.

This approach has the disadvantage of making `ModuleValue` more special.
But it has the advantage of not leaking the concept of properties deeper
into the type system.

suo committed:
Update on "[jit] fix @property"


Currently `@property` methods are broken in TorchScript because we
basically treat it as an attribute in the existing path: we'll evaluate
the method once and store that as the value forever.

This PR fixes it so that `@property` actually behaves correctly:
1. Added `properties` as part of `ConcreteModuleType`.
2. Added a `PropertyValue` `SugaredValue` that behaves exactly like a
method but it can be used as a value.
3. `ModuleValue::attr` now checks wheter a method is a property, and
returns a `PropertyValue` if so.

This approach has the disadvantage of making `ModuleValue` more special.
But it has the advantage of not leaking the concept of properties deeper
into the type system.

Differential Revision: [D18054946](https://our.internmc.facebook.com/intern/diff/D18054946)

suo committed:
Update on "[jit] fix @property"


Currently `@property` methods are broken in TorchScript because we
basically treat it as an attribute in the existing path: we'll evaluate
the method once and store that as the value forever.

This PR fixes it so that `@property` actually behaves correctly:
1. Added `properties` as part of `ConcreteModuleType`.
2. Added a `PropertyValue` `SugaredValue` that behaves exactly like a
method but it can be used as a value.
3. `ModuleValue::attr` now checks wheter a method is a property, and
returns a `PropertyValue` if so.

This approach has the disadvantage of making `ModuleValue` more special.
But it has the advantage of not leaking the concept of properties deeper
into the type system.

Differential Revision: [D18054946](https://our.internmc.facebook.com/intern/diff/D18054946)

suo committed:
Update on "[jit] fix @property"


Currently `@property` methods are broken in TorchScript because we
basically treat it as an attribute in the existing path: we'll evaluate
the method once and store that as the value forever.

This PR fixes it so that `@property` actually behaves correctly:
1. Added `properties` as part of `ConcreteModuleType`.
2. Added a `PropertyValue` `SugaredValue` that behaves exactly like a
method but it can be used as a value.
3. `ModuleValue::attr` now checks wheter a method is a property, and
returns a `PropertyValue` if so.

This approach has the disadvantage of making `ModuleValue` more special.
But it has the advantage of not leaking the concept of properties deeper
into the type system.

Differential Revision: [D18054946](https://our.internmc.facebook.com/intern/diff/D18054946)

zdevito commented on torch/jit/_recursive.py:
```diff
@@ -408,6 +395,8 @@ def init_fn(script_module):
     # original function => [(mangled overload name, overload function)]
     overloads = {}
+
+    # TODO: should this be dir(type(mod))?
```
probably?

suo committed:
Update on "[jit] fix @property"


Currently `@property` methods are broken in TorchScript because we
basically treat it as an attribute in the existing path: we'll evaluate
the method once and store that as the value forever.

Since lack of `@property` support is easily worked around (just make it
a method), I've opted to just explicitly error to avoid confusion. If
people want it, they can file an issue and we can look at their use
case.

This also helps us nicely clean up some parts of the ScriptModule conversion
path.

Pull Request resolved: https://github.com/pytorch/pytorch/pull/28395

Differential Revision: [D18054946](https://our.internmc.facebook.com/intern/diff/D18054946)

kostmo commented:
## CircleCI build failures summary
As of commit 985eecb:
* 0/2 recognized as flaky
* 2/2 broken upstream. You may want to rebase on the [latest viable branch](https://dr.pytorch.org/master-viable-commits.html).
* 0/2 failures introduced in this PR



Here are the [reasons each build failed](https://dr.pytorch.org/commit-details.html?sha1=985eecbfe2ff53488b6c6a42652cb7bd96c2afd4).

---
This comment was automatically generated by [Dr. CI](https://github.com/kostmo/circleci-failure-tracker/tree/master/docs/from-pull-request-comment).
Follow [this link to opt-out](https://dr.pytorch.org/admin/comments-opt-out.html) of these comments for your Pull Requests.


Please report bugs/suggestions on the [GitHub issue tracker](https://github.com/kostmo/circleci-failure-tracker/issues).

suo committed:
Update on "[jit] fix @property"


Currently `@property` methods are broken in TorchScript because we
basically treat it as an attribute in the existing path: we'll evaluate
the method once and store that as the value forever.

Since lack of `@property` support is easily worked around (just make it
a method), I've opted to just explicitly error to avoid confusion. If
people want it, they can file an issue and we can look at their use
case.

This also helps us nicely clean up some parts of the ScriptModule conversion
path.

Pull Request resolved: https://github.com/pytorch/pytorch/pull/28395

Differential Revision: [D18054946](https://our.internmc.facebook.com/intern/diff/D18054946)

suo committed:
Update on "[jit] fix @property"


Currently `@property` methods are broken in TorchScript because we
basically treat it as an attribute in the existing path: we'll evaluate
the method once and store that as the value forever.

Since lack of `@property` support is easily worked around (just make it
a method), I've opted to just explicitly error to avoid confusion. If
people want it, they can file an issue and we can look at their use
case.

This also helps us nicely clean up some parts of the ScriptModule conversion
path.

Pull Request resolved: https://github.com/pytorch/pytorch/pull/28395

Differential Revision: [D18054946](https://our.internmc.facebook.com/intern/diff/D18054946)
